pkg/pool/pool.go

```diff
@@ -108,28 +108,65 @@ func newPool(ctx context.Context, options *BuilderOptions) (Pool, error) {
 	ownerID := owner.NewIDFromNeo3Wallet(wallet)
 
 	pool := &pool{sampler: sampler, owner: ownerID, clientPacks: clientPacks}
-	go func() {
-		ticker := time.NewTimer(options.ClientRebalanceInterval)
-		for range ticker.C {
-			ok := true
-			for i, clientPack := range pool.clientPacks {
-				func() {
-					tctx, c := context.WithTimeout(ctx, options.NodeRequestTimeout)
-					defer c()
-					if _, err := clientPack.client.EndpointInfo(tctx); err != nil {
-						ok = false
-					}
-					pool.lock.Lock()
-					pool.clientPacks[i].healthy = ok
-					pool.lock.Unlock()
-				}()
-			}
-			ticker.Reset(options.ClientRebalanceInterval)
-		}
-	}()
+	go startRebalance(ctx, pool, options)
 	return pool, nil
 }
 
+func startRebalance(ctx context.Context, p *pool, options *BuilderOptions) {
+	ticker := time.NewTimer(options.ClientRebalanceInterval)
+	newWeights := make([]float64, len(options.weights))
+
+	for range ticker.C {
+		wg := sync.WaitGroup{}
+		wg.Add(len(p.clientPacks))
+		for i, cPack := range p.clientPacks {
+			go func(i int, netmap client.Netmap) {
+				ok := true
+				defer wg.Done()
+				tctx, c := context.WithTimeout(ctx, options.NodeRequestTimeout)
+				defer c()
+				if _, err := netmap.EndpointInfo(tctx); err != nil {
+					ok = false
+					newWeights[i] = 0
+				}
+				if ok {
+					newWeights[i] = options.weights[i]
+				}
+
+				p.lock.Lock()
+				p.clientPacks[i].healthy = ok
+				p.lock.Unlock()
+			}(i, cPack.client)
+		}
+		wg.Wait()
+
+		newWeights = adjustWeights(newWeights, options.weights)
+		source := rand.NewSource(time.Now().UnixNano())
+		p.lock.Lock()
+		p.sampler = NewSampler(newWeights, source)
```

Comment:
If weights are not changed (most of the time that's the case) we shouldn't be doing this.